packages/codemods/src/helpers/result.ts

```diff
@@ -0,0 +1,127 @@
+import { Chainable, Flattened } from './chainable';
+
+export interface Ok<R, E> extends ResultInternal<R, E> {
+  ok: true;
+  value: R;
+}
+
+export interface Err<R, E> extends ResultInternal<R, E> {
+  ok: false;
+  value: E;
+}
+
+class ResultInternal<R, E> implements Chainable<R> {
+  public ok: boolean;
+  public value: R | E;
+
+  public constructor(options: { ok: boolean; value: R | E }) {
+    this.ok = options.ok;
+    this.value = options.value;
+  }
+
+  public map<T>(this: Result<R, E>, fn: (v: R) => T): Result<T, E> {
+    if (this.ok) {
+      return Ok(fn(this.value));
+    }
+
+    return Err(this.value);
+  }
+
+  public chain<T>(this: Result<R, E>, fn: (v: R) => Result<T, E>): Result<T, E> {
+    if (this.ok) {
+      return fn(this.value);
+    }
+    return Err(this.value);
+  }
+
+  /**
+   * Works just like chain, but is only called if this Result is an error.
+   * This returns a new Result with type Result<R, NewType>
+   */
+  public orChain<T>(this: Result<R, E>, fn: (v: E) => Result<R, T>): Result<R, T> {
+    if (!this.ok) {
+      return fn(this.value);
+    }
+    return Ok(this.value);
+  }
+
+  public flatten(): Flattened<R, Result<R, E>> {
+    if (this.value && this.value instanceof ResultInternal) {
+      return this.value as Flattened<R, Result<R, E>>;
+    }
+    return (this as unknown) as Flattened<R, Result<R, E>>;
+  }
+
+  /**
+   * This allows users to opperate on a presumed result without needing to know whether or not
+   * the result was successful or not.
+   * At each call if it is a result type of ok, it will call the supplied function, otherwise it
+   * will return the current value with a new Err.
+   *
+   * @param fnOk Function that takes in an ok value of type R and returns either F or Result<F,E>
+   */
+  public then<F>(this: Result<R, E>, fnOk: (v: R) => F | Result<F, E>): Result<F, E> {
+    if (this.ok) {
+      return Ok(fnOk(this.value)).flatten() as Result<F, E>;
+    }
+
+    return Err(this.value);
+  }
+
+  /**
+   * Works just like then, but is only called if this Result is an error.
+   * This returns a new Result with type Result<R, NewType>
+   */
+  public orThen<F>(this: Result<R, E>, fnErr: (v: E) => F | Result<R, F>): Result<R, F> {
+    if (!this.ok) {
+      return Ok(fnErr(this.value)).flatten() as Result<R, F>;
```

Comment:
shouldn't this be Err?

Reply:
Yes it should be. Or at least I need to change the typings.